crates/cubecl-core/src/frontend/element/float/tensor_float.rs

```diff
@@ -0,0 +1,297 @@
+#[repr(transparent)]
+#[derive(Clone, Copy, Default, Serialize, Zeroable, Pod, PartialEq, PartialOrd)]
+pub struct tf32(u32);
+
+impl tf32 {
+    /// Constructs a [`tf32`] value from the raw bits.
+    #[inline]
+    #[must_use]
+    pub const fn from_bits(bits: u32) -> tf32 {
+        tf32(bits)
+    }
+
+    /// Constructs a [`tf32`] value from a 32-bit floating point value.
+    ///
+    /// This operation is lossy. If the 32-bit value is too large to fit, ±∞ will result. NaN values
+    /// are preserved. Subnormal values that are too tiny to be represented will result in ±0. All
+    /// other values are truncated and rounded to the nearest representable value.
+    #[inline]
+    #[must_use]
+    pub const fn from_f32(value: f32) -> tf32 {
+        tf32(f32_to_tf32(value))
+    }
+
+    /// Constructs a [`tf32`] value from a 64-bit floating point value.
+    ///
+    /// This operation is lossy. If the 64-bit value is to large to fit, ±∞ will result. NaN values
+    /// are preserved. 64-bit subnormal values are too tiny to be represented and result in ±0.
+    /// Exponents that underflow the minimum exponent will result in subnormals or ±0. All other
+    /// values are truncated and rounded to the nearest representable value.
+    #[inline]
+    #[must_use]
+    pub const fn from_f64(value: f64) -> tf32 {
+        tf32(f32_to_tf32(value as f32))
+    }
+
+    /// Converts a [`tf32`] into the underlying bit representation.
+    #[inline]
+    #[must_use]
+    pub const fn to_bits(self) -> u32 {
+        self.0
+    }
+
+    /// Converts a [`tf32`] value into an [`f32`] value.
+    ///
+    /// This conversion is lossless as all values can be represented exactly in [`f32`].
+    #[inline]
+    #[must_use]
+    pub const fn to_f32(self) -> f32 {
+        tf32_to_f32(self.0)
+    }
+
+    /// Converts a [`tf32`] value into an [`f64`] value.
+    ///
+    /// This conversion is lossless as all values can be represented exactly in [`f64`].
+    #[inline]
+    #[must_use]
+    pub const fn to_f64(self) -> f64 {
+        tf32_to_f32(self.0) as f64
+    }
+}
+
+#[inline]
+pub(crate) const fn f32_to_tf32(value: f32) -> u32 {
+    // TODO: Replace mem::transmute with to_bits() once to_bits is const-stabilized
+    // Convert to raw bytes
+    let x: u32 = unsafe { mem::transmute::<f32, u32>(value) };
+
+    // check for NaN
+    if x & 0x7FFF_FFFFu32 > 0x7F80_0000u32 {
+        // Keep high part of current mantissa but also set most significiant mantissa bit
+        return (x >> 13) | (0x0200u32);
+    }
+
+    x >> 13
+}
+
+#[inline]
+pub(crate) const fn tf32_to_f32(i: u32) -> f32 {
+    // TODO: Replace mem::transmute with from_bits() once from_bits is const-stabilized
+    // If NaN, keep current mantissa but also set most significiant mantissa bit
+    if i & 0x7FFFu32 > 0x7F80u32 {
+        unsafe { mem::transmute::<u32, f32>((i | 0x0200u32) << 13) }
+    } else {
+        unsafe { mem::transmute::<u32, f32>(i << 13) }
+    }
```

Comment:
Is there some tests for that?

Reply:
I actually need to change that, from testing it seems to actually have a compatible layout to `f32`, it's not shifted